time/date.jsonnet

```diff
@@ -34,8 +34,52 @@ local months_seconds(year, month) = std.foldl(
 // days_seconds returns the number of seconds in all days up to day-1.
 local days_seconds(day) = (day - 1) * 24 * 3600;
 
+// is_numeric returns a bool indicating whether the input string is numeric
+local is_numeric(input) = std.length(input) > 0 && std.length(std.prune([
+  if std.codepoint('0') > std.codepoint(char) || std.codepoint(char) > std.codepoint('9') then char
+  for char in std.stringChars(input)
+])) == 0;
+
+// parse_date_or_time parses input which has part_names separated by sep.
+local parse_date_or_time(input, sep, part_names) = (
+  local parts = std.split(input, sep);
+  assert std.length(parts) == std.length(part_names) : 'expected %(expected)d parts separated by %(sep)s in %(format)s formatted input "%(input)s", but got %(got)d' % {
+    expected: std.length(part_names),
+    sep: sep,
+    format: std.join(sep, part_names),
+    input: input,
+    got: std.length(parts),
+  };
+
+  {
+    [part_names[i]]:
+      // Fail with meaningful message if not numeric, otherwise it will be a hell to debug.
+      assert is_numeric(parts[i]) : '%(part_name)%s part "%(part)s" of %(format)s of input "%(input)s" is not numeric' % {
+        part_name: part_names[i],
+        part: parts[i],
+        format: std.join(sep, part_names),
+        input: input,
+      };
+      std.parseInt(parts[i])
+    for i in std.range(0, std.length(parts) - 1)
+  }
+);
+
 {
-  // date_to_unix_timestamp transforms a date into a unix timestamp.
+  // to_unix_timestamp transforms a date into a unix timestamp.
   to_unix_timestamp(year, month, day, hour, minute, second)::
     years_seconds(year) + months_seconds(year, month) + days_seconds(day) + hour * 3600 + minute * 60 + second,
+
+  // parse_rfc3339 parses an RFC3339 timestamp in UTC (ending in 'Z') with date and time separated by 'T', like '2006-01-02T15:04:05Z'
+  // The returned object has "year", "month", "day", "hour", "minute" and "second" keys, as well as a to_unix_timestamp method that returns the unix timestamp of the parsed date.
+  parse_rfc3339(input)::
+    assert std.endsWith(input, 'Z') : 'the provided RFC3339 "%s" should end with "Z"' % input;
```

Comment:
According to the RFC, this can also be `z`

> NOTE: Per [ABNF] and ISO8601, the "T" and "Z" characters in this syntax may alternatively be lower case "t" or "z" respectively.
      
https://www.ietf.org/rfc/rfc3339.txt

However, it also says that it is valid to restrict the set to just the uppercase so it's up to you if you want to support both.